### ISSUE TITLE
Plane suddenly going to steep climb during Autopilot mode.

### ISSUE THREAD

**Issue by eaglerainbow** [it0uchpods/IDG-A32X#86]
### Environment
Flightgear version 2018.3.2
IDG-A32X as of github commit 23d3576bf5b88c72bc66f496434bec5ab9b94fa7

### Background
I was having a flight from ELLX to EDDF and during taxi after arrival at EDDF I had the strange situation that my aircraft simply did not react to any input anymore. 
A fellow sim-airmen reported that he also had some issues with the most current master version (0c5be65001111d149f313607e8fc20d551bd7f3b) and suggested to go back a little. I chose 23d3576bf5b88c72bc66f496434bec5ab9b94fa7. I restarted the simulator completely and departed from EDDF heading to EDDP using autopilot (multiplayer was connected). The journey was uneventful until around 50 NM to EDDP, where I encountered a sudden very strange behavior of the airplane.

### Current Behavior
See [flight recording](https://github.com/it0uchpods/IDG-A32X/files/3026028/A320-200-IAE-20190330-231946.zip) at time index 47:45. In a nutshell (most of the stuff from my memory):
* AP1 and Autothrust engaged, running on Flight Director; route is entered properly; plane is right on the route for more than the last 10 minutes.
* The plane suddenly started to climb sharply. As a consequence autothrust increased throttle to MAX.
* After a couple of seconds of climb, plane lost thrust on both engines - engines went to state "off" even though engine master switches were set to "on". Engine did not spin down, but N1 and FF was suddenly "XX".
* RAT was deployed automatically.
* Autopilot was disengaged with master alert.
* I don't think I ever reached a stalling airspeed (but the AoA was unusually significant for sure).
* Auto-trimming still worked fine for a certain time, but failed shortly thereafter (manual trim message on the PFD).
* The last state of auto-trimming (which I only checked quite late) was roughly 3° nose up (so, either the issue was introduced by trimming, or auto-trimming still worked to a certain degree, because "3° nose up" wasn't the worst setting for gliding...)

By starting the APU, I was able to restart both engines (engine 2 went first) using APU bleed (I did not dare to test out a windmill restart). Auto-trimming did not come back anymore, also not the autopilots (the latter sounds logical to me, if auto-trimming is gone).

(Please ignore my shameful bad landing at the end, but I was confused by the fact that the runway wasn't illuminated and also still thrilled upon the recent surprise).

### Expected Behavior
Plane flights straight if autopilot is engaged and everything's on route... (and as long as I don't trigger a "catastrophe" manually, which I didn't). 

### Possible Solution
Human error by airmen?

### Steps to Reproduce
Sorry, I have no clue what might have triggered that.
Do you have a DFDR on board? 😀 If yes, I'd hope you could give me a hint what went wrong. 

#### Post Scriptum
Chapeau for the quality of your implementation! During the abnormal situation, I couldn't make out any physically irregular behavior or unlogical airplane system behavior.

**Comment by it0uchpods** [comment]
Hi.
I recommend NEVER downgrading the aircraft, unless you DELETE your IDG-A32X-config.xml in $FGHOME/Export. Possibly it messed up something in the aircraft. It would be better to report the issues, that we can fix them.

Anyways, that sounds very odd. The only thing I can think of is that the Weather thing caused a rapid increase of airspeed do to it jumping properly, which oversped the aircraft, causing the A/P to trip off, and the FlyByWire to nose up to achieve a DECEL rate.

I have NO idea why the engines would have failed. Were the fuel pumps on? If they were off, low G could cause gravity feed to fail, and they will die.

Sorry, the recorders are useless. We do not waste CPU or RAM by recording all the internal system states. So I will need a reproducible scenario in order to debug it.

If you encounter issues, please screenshot OFTEN whenever the behavior changes, so that I can have more clue of what is going on.
Dumping the property tree may help too, (debug menu)

If you encounter it again, please do those 2 steps, and let me know! Thanks!

And I'm glad you like them!

Kind Regards,
Josh

**Comment by eaglerainbow** [comment]
Hi,

First of all many thanks for your quick response!

> I recommend NEVER downgrading the aircraft, unless you DELETE your IDG-A32X-config.xml in $FGHOME/Export.

That'll be yet another thing, which would IMO worth to mention in an FAQ...

If you are interested, here is my `IDG-A32X-config.xml`:
```xml
<?xml version="1.0"?>

<PropertyList>
  <revision type="double">4785</revision>
  <keyboard-mode type="double">0</keyboard-mode>
  <laptop-mode type="double">0</laptop-mode>
  <adirs-skip type="double">1</adirs-skip>
  <welcome-skip type="double">0</welcome-skip>
  <no-rendering-warn type="double">0</no-rendering-warn>
  <pfd-rate type="double">1</pfd-rate>
  <nd-rate type="double">1</nd-rate>
  <uecam-rate type="double">1</uecam-rate>
  <lecam-rate type="double">1</lecam-rate>
  <iesi-rate type="double">1</iesi-rate>
  <autopush>
    <show-route type="double">1</show-route>
    <show-wingtip type="double">1</show-wingtip>
  </autopush>
</PropertyList> 
```

(BTW: I see that you have the highest revision in your export config XML. IIRC you show a message box during boot up, if the revisions do not match. Would it possible to show a similar message box on start-up, if the export config XML's revision was higher than the aircraft revision installed, warning about the downgrade? Just an idea...)

> It would be better to report the issues, that we can fix them.

I agree. If I see "LH2022" on the multiplayer again, I'll let him know that he should at least report his issue here (I don't know how reproducible his problem is).

> The only thing I can think of is that the Weather thing

If it helps, I could reconstruct the METAR, which I used during that flight:
``` 
EDDF 301920Z 24004KT CAVOK 14/02 Q1020 NOSIG 
```
(BTW: That's the real-world weather only copied, as my version still suffers from the know weather fetching bug). "Advanced weather modelling" was enabled.

If it helps, I had a similar "experience" some weeks back (which might also take the heat off your IDG-A320 😀 ): I was flying somewhere over the [Hunsrück](https://en.wikipedia.org/wiki/Hunsr%C3%BCck) in a C172P heading NW. Autopilot was engaged and it became quite boring. Therefore, I increased simulation speed to 8x. Out of a clear blue sky, I crashed so fast to ground that I couldn't understand what was going on. I triggered instant replay and found that due to an unknown reason to me (I did not touch any flight control), the vertical speed of the plane started to (IIRC) jump around +/- 500 ft/min and suddenly took off into the positive scale. Values became too large for the meter to be shown properly anymore. There is no auto-throttling in the Cessna, so the plane started to stall and went into tailspin. During falling the IAS became too high, so several components (e.g. wings) broke. As simulation speed was high, I had no chance to recover the plane. 
If you are interested in this, I might still have the tape recording of it somewhere around (would have to look for it).

> caused a rapid increase of airspeed do to it jumping properly, which oversped the aircraft, causing the A/P to trip off, and the FlyByWire to nose up to achieve a DECEL rate.

Well, I can't really believe in that theory: If that had been the case, then the A/P wouldn't have throttled up the engines when the AoA increased - but that was clearly the case. It had been the sound of the engines roaring up, which alerted me in first place.

> I have NO idea why the engines would have failed. Were the fuel pumps on? If they were off, low G could cause gravity feed to fail, and they will die.

To be honest, I don't remember. Moreover, I have to admit that I often tend to forget switching them on.
Thus, your "low G" idea sounds reasonable to me. Perhaps this was just a "second unfortunate event" to happen as a consequence out of the first one. If I find time today or tomorrow, I'll give it a try and will report back the results here.

> Sorry, the recorders are useless.

Okay, I didn't know that. Yet another thing which might be worth an FAQ item 😄 

> So I will need a reproducible scenario in order to debug it.

Many thanks for even trying that!
I fear this won't be an easy task: I have been flying several hours with it, but did not encounter anything faintly similar. There must be some sort of "external factor" with rare probability, which triggers the issue.

> If you encounter issues, please screenshot OFTEN whenever the behavior changes, so that I can have more clue of what is going on. Dumping the property tree may help too, (debug menu)

And even yet another statement worth for the FAQ 😃

**Comment by it0uchpods** [comment]
Hi,
Thanks for your comments. Yes, that is possible. The existing code supports doing that, I just need to add another check. Or, I could just kill and reset the XML from acconfig's files and then warn the users the settings might have been reset.

That sounds like it might be why. 8x is very fast, if you had a long lagspike, it may have caused bad things to happen. The autopilot and FlyByWire in the Airbus are very complex, so if some components did not run, the system would get a bit messed up. I recommend never going above 4x unless you can do so while maintain a FPS of at least 10. (I recommend an FPS of 10 as the minimum amount for any of my planes to fly correctly while on A/P just in case too)

Yes, a FAQ is long overdue, and I've been asked to add one too. I will be in contact shortly with some more things, if you don't mind adding them? (My write skills leave some to be desired) Thank you!

Kind Regards,
Josh

**Comment by eaglerainbow** [comment]
> That sounds like it might be why. 8x is very fast,

Note that I did not change the simulation speed in your A320 case (it was only with the C172P).

>  I recommend never going above 4x unless you can do so while maintain a FPS of at least 10.

This also tastes like another entry for the FAQ 😀 

> I will be in contact shortly with some more things, if you don't mind adding them?

I'd love to! Please help yourself!

**Comment by it0uchpods** [comment]
> Note that I did not change the simulation speed in your A320 case (it was only with the C172P).

Crap, and I thought I was on to something! :D

OK, I'll have to think on this a bit.

Will let you know regarding FAQ

Kind Regards,
Josh

**Comment by eaglerainbow** [comment]
> Will let you know regarding FAQ

If I will be able to understand what the FAQ's entry is about, and if you *want*, I can also have a look at your topic. As long it's just writing text, I can try to be of (a little) help 😄

**Comment by eaglerainbow** [comment]
>> I recommend never going above 4x unless you can do so while maintain a FPS of at least 10.

> This also tastes like another entry for the FAQ 😀

Just for the record: I have enhanced the PR for the FAQ on this with 9f6f084

**Comment by it0uchpods** [comment]
Sounds good.
You may want to revise showing frame rate, it is enable in the View -> View Options -> Show frame rate :)

I will look thru this later, going to do some flying today.
J

**Comment by eaglerainbow** [comment]
> You may want to revise showing frame rate, it is enable in the View -> View Options -> Show frame rate :)

Very valid feedback - fixed with bbf03d8

> I will look thru this later, going to do some flying today.

Have fun - and think about this issue here! 😁

**Comment by it0uchpods** [comment]
thanks!

**Comment by eaglerainbow** [comment]
*Please don't read all this today/tonight anymore - please keep on playing!*

Trying things out a little, I think I found some answers:

## Initial state of scenario

First of all, here's always the starting point for my scenario: 
* Initialized using A/C with "ready to take-off"
* Having started from the runway, keeping heading of the runway. 
* Climbing to FL100 using AP1 engaged. 
* Waiting until level flight.

![Initial state](https://user-images.githubusercontent.com/12811079/55295187-86d4a600-540a-11e9-87c4-f37269e5df55.png)

As a result, trim is roughly 2.5° nose up.

## 1st variant
* Fuel pumps running
* Very harsh pulling of the yoke for several seconds up to the maximum of the axis.
* Then pushing back to get back to level flight

### Result:
* AP is dis-engaged with master warning (as expected)
* Trim does not change.
* No auto-trim failure
* Engine keep on running even if crossing V/S to negative scale.

## 2nd variant
* Fuel pumps off
* Very harsh pulling of the yoke for several seconds up to the maximum of the axis.
* Then pushing back to get back to level flight

### Result:
* AP is dis-engaged with master warning (as expected)
* Trim does change (more nose up)
* No auto-trim failure
* Engine die shortly after crossing V/S into negative scale.

## 3rd variant
* Fuel pumps running
* Moving the mouse down quite harshly in flight control mode
* Then moving the mouse back up roughly the same way

### Result:
* AP is dis-engaged with master warning (as expected)
* Trim does change significantly (much more nose up)
* No auto-trim failure
* Engines keep running


## 4th variant
* Fuel pumps disengaged
* Moving the mouse down quite harshly in flight control mode
* Then moving the mouse back up roughly the same way

### Result:
* AP is dis-engaged with master warning (as expected)
* Trim does change significantly (3.5° nose up)
* auto-trim failure triggered
* Engines die shortly after crossing V/S into negative scale.
* Even after engines are restored, it's not possible to reengage the AP anymore (auto-thrust is still available though). 
* AP can not be reengaged even if trim is corrected manually.
* Trim behavior is significantly different: If I go back to 2.5° nose up, then the plane is trimmed significantly too much "nose down"

![4th variant screenshot](https://user-images.githubusercontent.com/12811079/55295211-c8fde780-540a-11e9-9343-1da174b3dbf9.png)


## My conclusion(s)
Here is what I make out of this:
* Especially the 4th variant felt very much like the situation I was in on reporting this issue here.
* Your assumption about zero gravity and fuel pumps seems to be correct.  I must have flown with the fuel pumps disengaged (which, as said, does not surprise me much).
* Simply pulling the yoke is not sufficient to trigger the issue. The mouse is needed to get the same effect.

I *believe* that I did not touch the mouse when the issue reported initially happened. However, I also wouldn't take an oath on that either.

In short:
* It was either me (pilot error), or
* the effect must have been very harsh - so harsh that you cannot trigger it with the yoke, but you need to have the mouse in flight mode.

Even if it was me, I still struggle a little with an explanation on the totally different trim values. Or does the RAT have such a big influence on trim?

**Comment by it0uchpods** [comment]
Hi,
It is not negative V/S, but negative Gload. Since you push down, which causes negative G, this mostly makes V/S less, which may make it appear like negative V/S is causing it.

1st Variant:
TRIM does not change in high Gload IIRC, so that makes sense. Everything here looks right.

2nd Variant:
I am guessing you did not hydraulic pressure, so the FlyByWire stays in NORMAL law, hence why trim is normal

3rd Variant:
Normal

4th Variant:
This time you lost hydraulic pressure breifly, or another thing caused the FlyByWire to go to DIRECT law, which means no control law. Inputs are passed directly to the surfaces, and autotrim does not function. See the "USE MAN PITCH TRIM" in amber on the PFD.

Engines are mounted below the wing, so they cause an upwards pitching moment as thrust increase. So as power changes, you need to adjust trim. Usually, the FlyByWire Control in NORMAL and ALTERNATE law uses CStar Logic, this is G and pitch rate control blended. Meaning, at 0 stick, pitch attitude remain constant. The autotrim trims to keep the FlyByWire elevator demand neutral. in DIRECT law, since there is no control law, you need to compensate trim yourself.

This all seems right. The only I see here is the inconsistent behavior of the hydraulic or FlyByWire systems when engines fail. I will do some tests sometimes, but I'm going to do a few tests later.

Kind Regards,
Josh

**Comment by eaglerainbow** [comment]
> I am guessing you did not hydraulic pressure, 

What do you mean by that? Is there another parameter which could influence the behavior I don't know yet?

> Inputs are passed directly to the surfaces, and autotrim does not function. [...] in DIRECT law, since there is no control law, you need to compensate trim yourself.

That starts to explain quite a lot to me - so we get the difference in the trimming behavior I couldn't understand yet (well, you see that I am not an educated, professional pilot, but only a flight beginner -- thanks for all your explanations&support!).

> The only I see here is the inconsistent behavior of the hydraulic or FlyByWire systems when engines fail.

What is exactly what causes here your headache? Can I be of help somehow performing some tests and reporting the results?

**Comment by eaglerainbow** [comment]
Interesting to know!

> If pressure was lost momentarily, that may have happened.

Could it be that zero G cause the pressure to drop significantly enough?
If that was the case, though, how can an [A340 fly with zero G](https://www.flightglobal.com/news/articles/hi-fly-supplies-a340-for-commercial-zero-gravity-fli-424119/)?

**Comment by it0uchpods** [comment]
Closing, will make a new issue to better address the fault.
